include/picolibrary/wiznet/w5500/ip/tcp.h

```diff
@@ -23,10 +23,1320 @@
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            switch ( static_cast<Buffer_Size>( result.value() ) ) {
+                case Buffer_Size::_1_KIB: buffer_size = 1 * 1024; break;
+                case Buffer_Size::_2_KIB: buffer_size = 2 * 1024; break;
+                case Buffer_Size::_4_KIB: buffer_size = 4 * 1024; break;
+                case Buffer_Size::_8_KIB: buffer_size = 8 * 1024; break;
+                case Buffer_Size::_16_KIB: buffer_size = 16 * 1024; break;
+                default: return m_network_stack->nonresponsive_device_error();
+            } // switch
+        }
+
+        Size free_size;
+        {
+            auto result = m_driver->read_sn_tx_fsr( m_socket_id );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            free_size = result.value();
+        }
+
+        if ( free_size > buffer_size ) {
+            return m_network_stack->nonresponsive_device_error();
+        } // if
+
+        return static_cast<Size>( buffer_size - free_size );
+    }
+
+    /**
+     * \brief Transmit data to the remote endpoint.
+     *
+     * \param[in] begin The beginning of the block of data to write to the socket's
+     *            transmit buffer.
+     * \param[in] end The end of the block of data to write to the socket's transmit
+     *            buffer.
+     *
+     * \return The end of the data that was written to the socket's transmit buffer if
+     *         writing data to the socket's transmit buffer succeeded.
+     * \return picolibrary::Generic_Error::NOT_CONNECTED if the socket is not connected to
+     *         a remote endpoint.
+     * \return picolibrary::Generic_Error::WOULD_BLOCK if no data could be written to the
+     *         socket's transmit buffer without blocking.
+     * \return picolibrary::WIZnet::W5500::IP::Network_Stack::nonresponsive_device_error()
+     *         if the W5500 is nonresponsive.
+     * \return An error code if writing data to the socket's transmit buffer failed for
+     *         any other reason.
+     */
+    auto transmit( std::uint8_t const * begin, std::uint8_t const * end ) noexcept
+        -> Result<std::uint8_t const *, Error_Code>
+    {
+        // #lizard forgives the length
+        // #lizard forgives the complexity
+
+        if ( m_state != State::CONNECTED ) {
+            return Generic_Error::NOT_CONNECTED;
+        } // if
+
+        {
+            auto result = m_driver->read_sn_sr( m_socket_id );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            switch ( static_cast<Socket_Status>( result.value() ) ) {
+                case Socket_Status::CLOSED: return Generic_Error::NOT_CONNECTED;
+                case Socket_Status::ESTABLISHED: break;
+                case Socket_Status::FIN_WAIT: return Generic_Error::WOULD_BLOCK;
```

Comment:
This function should not be called once a graceful shutdown is in progress. Return `Generic_Error::NOT_CONNECTED` instead of `Generic_Error::WOULD_BLOCK` for this socket status for better consistency.